File: backend/app/schemas/student.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class StudentCreate(BaseModel):
    name: str
    board: str = Field(default="cbse", pattern=r"^(cbse|icse|ib)$")
    grade: int = Field(ge=4, le=9)
    grade_band: str = Field(pattern=r"^(4-5|6-7|8-9)$")
    cohort_id: uuid.UUID | None = None


class StudentOut(BaseModel):
    id: uuid.UUID
    name: str
    board: str
    grade: int
    grade_band: str
    cohort_id: uuid.UUID | None = None
    diagnostic_completed: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class CompetencyStateOut(BaseModel):
    competency_id: str
    p_learned: float
    p_transit: float
    p_guess: float
    p_slip: float
    total_evidence: int
    consecutive_failures: int
    is_stuck: bool
    last_evidence_at: datetime | None = None
    stability: float
    avg_response_time_ms: float | None = None
    stage: int
    confidence: float
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class StudentStateOut(BaseModel):
    student: StudentOut
    states: list[CompetencyStateOut]

File: frontend/src/api/types/student.ts
export interface Student {
  id: string;
  name: string;
  board: "cbse" | "icse" | "ib";
  grade: number;
  grade_band: "4-5" | "6-7" | "8-9";
  cohort_id: string | null;
  diagnostic_completed: boolean;
  created_at: string;
}

export interface CompetencyState {
  competency_id: string;
  p_learned: number;
  p_transit: number;
  p_guess: number;
  p_slip: number;
  total_evidence: number;
  consecutive_failures: number;
  is_stuck: boolean;
  last_evidence_at: string | null;
  stability: number;
  avg_response_time_ms: number | null;
  stage: number;
  confidence: number;
  updated_at: string | null;
}

export interface StudentState {
  student: Student;
  states: CompetencyState[];
}
